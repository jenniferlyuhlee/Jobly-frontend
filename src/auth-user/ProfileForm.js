import React, { useState, useContext } from "react";
import UserContext from "./UserContext";
import JoblyApi from "../api";
import Alert from "../shared/Alert";
import UserJobs from "./UserJobs";

/**
 * Profile Component - shows form to edit profile info
 * State: formData, formErros, saveConfirmed, showProfile
 * Inherits currUser data and state from parent to make user changes
 * Inherits logout function to logout when account deleted
 * Toggles display between profile and jobs user applied to, UserJobs component
 */

function ProfileForm(){
    const {currUser, setCurrUser, logout} = useContext(UserContext);

    // states: formData, formErrors, saveConfirmed, showProfile, showAppliedJobs
    const [formData, setFormData] = useState({
        username: currUser.username,
        firstName: currUser.firstName,
        lastName: currUser.lastName,
        email: currUser.email,
    });
    const [formErrors, setFormErrors] = useState([])
    const [saveConfirmed, setSaveConfirmed] = useState(false);
    const [settingErrors, setSettingErrors] = useState(null);
    const [showProfile, setShowProfile] = useState(true);
    
    // general callback to update targeted form data
    function handleChange (e) {
        const { name, value } = e.target;
        setFormData(data => ({...data, [name]: value}));
        setFormErrors([]);
    }

    // form submission handler
    async function handleSubmit(e){
        e.preventDefault();

        const profileData = {
            firstName: formData.firstName,
            lastName: formData.lastName,
            email: formData.email,
        }
        let updatedUser;

        try{
            updatedUser=await JoblyApi.updateProfile(formData.username, profileData);
        }
        catch(err){
            setFormErrors(err)
            return;
        }

        // resets profile form to display updated info
        setFormData(f => ({...f}));
        setFormErrors([]);
        setSaveConfirmed(true);
        setCurrUser(updatedUser);
    }

    // delete user handler
    async function handleDelete(){
        try{
            const deleted = await JoblyApi.deleteUser(currUser.username)
            if (deleted){
                logout()
            }
        }
        catch(err){
            setSettingErrors(err)
            return 
        }
    }

    // toggle profile/applied jobs display 
    function toggleDisplay(){
        setShowProfile(state => !state)
    }

        return(
            <>
            <div className="d-flex justify-content-around m-3">
                <button onClick={toggleDisplay} 
                className={`btn btn-tab ${showProfile? "tertiary active-tab": "primary"}`}>
                    My Profile
                </button>
            
                <button onClick={toggleDisplay} 
                className={`btn btn-tab ${showProfile? "primary": "tertiary active-tab"}`}>
                    My Jobs
                </button>
            </div>
            {showProfile ? 
            <div className="SignupForm card p-4 my-3 col-lg-10 mx-auto">
                {/* Profile Info Update Form */}
                <h3 className="text-center mb-3">Edit Profile</h3>
                <form onSubmit={handleSubmit}>
                    <div className="row mb-3">
                        <label className="col-sm-3 col-form-label">
                            Username
                        </label>
                        <div className="col-sm-9 ">
                            <input
                                name="username"
                                className="form-control"
                                value={formData.username}
                                readOnly
                                disabled
                            />
                        </div>
                    </div>
                    <div className="row mb-3">
                        <label className="col-sm-3 col-form-label">
                            First Name
                        </label>
                        <div className="col-sm-9">
                            <input
                                name="firstName"
                                className="form-control"
                                value={formData.firstName}
                                onChange={handleChange}
                            />
                        </div>
                    </div>
                    <div className="row mb-3">
                        <label className="col-sm-3 col-form-label">
                            Last Name
                        </label>
                        <div className="col-sm-9">
                            <input
                                name="lastName"
                                className="form-control"
                                value={formData.lastName}
                                onChange={handleChange}
                            />
                        </div>
                    </div>
                    <div className="row mb-3">
                        <label className="col-sm-3 col-form-label">
                            Email
                        </label>
                        <div className="col-sm-9">
                            <input
                                name="email"
                                type="email"
                                className="form-control"
                                value={formData.email}
                                onChange={handleChange}
                            />
                        </div>
                    </div>
                    {formErrors.length>0?
                        <Alert type="danger" messages={formErrors} />
                        :
                        null
                    }
                    {saveConfirmed?
                        <Alert type="success" messages={["Updated successfully."]} />
                        :
                        null
                    }
                    <div className="text-center">
                        <button className="btn btn-success">Save Changes</button>
                    </div>
                </form>
                {/* Profile Settings */}
                <h3 className="m-3 text-center">Settings</h3>
                {settingErrors? 
                    <Alert type="warning" messages={["Unable to proceed with your request. Try again later."]} />
                    :
                    null}
                <div className="text-center">
                    <button onClick={handleDelete} className="btn btn-danger">
                        Delete Account
                    </button>
                </div>
            </div>
            : 
            <UserJobs />
            }
            </>
        )
}

export default ProfileForm;