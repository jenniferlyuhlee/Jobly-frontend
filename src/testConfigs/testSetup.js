import React from "react";
import UserContext from "../auth-user/UserContext";

const demoUser = {
    username: "testuser",
    first_name: "testfirst",
    last_name: "testlast",
    email: "test@test.com",
    photo_url: null,
    applications: []
  };

  const UserProvider =
      ({ children, currUser = demoUser, hasAppliedToJob = () => false }) => (
      <UserContext.Provider value={{ currUser, hasAppliedToJob }}>
        {children}
      </UserContext.Provider>
  );

  export {UserProvider};
