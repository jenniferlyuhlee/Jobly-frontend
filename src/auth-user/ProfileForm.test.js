import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { UserProvider } from "../testConfigs/testSetup";
import { MemoryRouter } from "react-router-dom";
import ProfileForm from "./ProfileForm";
import JoblyApi from "../api";

// Mock Jobly API
jest.mock("../api")

it("renders without crashing", function () {
    render(
        <MemoryRouter>
            <UserProvider>
            <ProfileForm />
          </UserProvider>,
        </MemoryRouter>)
})

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
        <UserProvider>
        <ProfileForm />
      </UserProvider>,
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("updates currUser's info when submitted", function () {
    JoblyApi.updateProfile.mockResolvedValue({
        username: "testuser",
        firstName: "newtestfirst",
        lastName: "newtestlast",
        email: "newemailemail.com"
    })

    const { getByLabelText, queryByText } = render(
        <MemoryRouter>
            <UserProvider>
            <ProfileForm />
          </UserProvider>,
        </MemoryRouter>
    );

    // grab form fields and submit button
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email");
    const btn = queryByText("Save Changes")

    // simulate user input and form submission
    fireEvent.change(firstNameInput, {target: {value: "newtestfirst"}});
    fireEvent.change(lastNameInput, {target: {value: "newtestlast"}});
    fireEvent.change(emailInput, {target: {value: "newemail@email.com"}});
    fireEvent.click(btn);

    expect(firstNameInput.value).toBe("newtestfirst")
    expect(lastNameInput.value).toBe("newtestlast")
    expect(emailInput.value).toBe("newemail@email.com")
});
