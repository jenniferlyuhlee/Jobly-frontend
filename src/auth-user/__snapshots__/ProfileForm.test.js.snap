// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-around m-3"
  >
    <button
      class="btn btn-tab tertiary active-tab"
    >
      My Profile
    </button>
    <button
      class="btn btn-tab primary"
    >
      My Jobs
    </button>
  </div>
  <div
    class="SignupForm card p-4 my-3 col-lg-10 mx-auto"
  >
    <h3
      class="text-center mb-3"
    >
      Edit Profile
    </h3>
    <form>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-3 col-form-label"
        >
          Username
        </label>
        <div
          class="col-sm-9 "
        >
          <input
            class="form-control"
            disabled=""
            name="username"
            readonly=""
            value="testuser"
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-3 col-form-label"
        >
          First Name
        </label>
        <div
          class="col-sm-9"
        >
          <input
            class="form-control"
            name="firstName"
            value=""
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-3 col-form-label"
        >
          Last Name
        </label>
        <div
          class="col-sm-9"
        >
          <input
            class="form-control"
            name="lastName"
            value=""
          />
        </div>
      </div>
      <div
        class="row mb-3"
      >
        <label
          class="col-sm-3 col-form-label"
        >
          Email
        </label>
        <div
          class="col-sm-9"
        >
          <input
            class="form-control"
            name="email"
            type="email"
            value="test@test.com"
          />
        </div>
      </div>
      <div
        class="text-center"
      >
        <button
          class="btn btn-success"
        >
          Save Changes
        </button>
      </div>
    </form>
    <h3
      class="m-3 text-center"
    >
      Settings
    </h3>
    <div
      class="text-center"
    >
      <button
        class="btn btn-danger"
      >
        Delete Account
      </button>
    </div>
  </div>
  ,
</DocumentFragment>
`;
